In [2]:
import pandas as pd
import numpy as np

## Load Dataset

In [3]:
df=pd.read_csv(r"/Users/prasannakumarreddy/Downloads/restaurant_sales_dataset.csv")

In [4]:
df

Order_ID        Date Restaurant_Name     Location Item_Category  \
0            1  2023-01-01     Grill House      Chicago         Pizza   
1            2  2023-01-02     Urban Bites          NaN         Salad   
2            3  2023-01-03  Spice Junction          NaN        Burger   
3            4  2023-01-04     Urban Bites      Phoenix           NaN   
4            5  2023-01-05     Urban Bites  Los Angeles        Burger   
...        ...         ...             ...          ...           ...   
2995      2996  2031-03-15     Grill House  Los Angeles         Pasta   
2996      2997  2031-03-16             NaN     New York         Salad   
2997      2998  2031-03-17  Spice Junction     New York         Pasta   
2998      2999  2031-03-18    Tasty Treats      Chicago         Pasta   
2999      3000  2031-03-19             NaN      Chicago           NaN   

      Quantity  Price_per_Item  Total_Sales Payment_Method  Customer_Age  \
0            3           30.46        91.38    Credit Card          57.0   
1            8           21.45          NaN    Credit Card          24.0   
2            8           47.13       377.04           Cash          22.0   
3            2            7.47        14.94     Debit Card          64.0   
4            4           35.30       141.20            UPI          59.0   
...        ...             ...          ...            ...           ...   
2995         1           33.05          NaN           Cash          39.0   
2996         8           34.98       279.84           Cash          34.0   
2997         2           29.59        59.18    Credit Card          55.0   
2998         1           14.24        14.24           Cash          42.0   
2999         8           38.98       311.84     Debit Card          49.0   

     Customer_Gender Delivery_Status  
0              Other             NaN  
1                NaN             NaN  
2             Female             NaN  
3             Female       Delivered  
4             Female             NaN  
...              ...             ...  
2995            Male       Cancelled  
2996             NaN       Delivered  
2997          Female       Cancelled  
2998           Other         Pending  
2999           Other         Pending  

[3000 rows x 12 columns]

### Observation:
 - Dataset has been loaded successfully using pd.read_csv()

### Understanding the Structure of the Dataset

In [5]:
df.shape

(3000, 12)

## Observation:
*  The dataset contains 3000 rows and 12 columns representing various sales of items in restaurant.

## previewing the first few rows

In [6]:
df.head()

Order_ID        Date Restaurant_Name     Location Item_Category  Quantity  \
0         1  2023-01-01     Grill House      Chicago         Pizza         3   
1         2  2023-01-02     Urban Bites          NaN         Salad         8   
2         3  2023-01-03  Spice Junction          NaN        Burger         8   
3         4  2023-01-04     Urban Bites      Phoenix           NaN         2   
4         5  2023-01-05     Urban Bites  Los Angeles        Burger         4   

   Price_per_Item  Total_Sales Payment_Method  Customer_Age Customer_Gender  \
0           30.46        91.38    Credit Card          57.0           Other   
1           21.45          NaN    Credit Card          24.0             NaN   
2           47.13       377.04           Cash          22.0          Female   
3            7.47        14.94     Debit Card          64.0          Female   
4           35.30       141.20            UPI          59.0          Female   

  Delivery_Status  
0             NaN  
1             NaN  
2             NaN  
3       Delivered  
4             NaN

### Observation:
 - The preview shows the first few rows of the data and helps us check if the file loaded correctly.

### Dataset Structure and Data Types


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Order_ID         3000 non-null   int64  
 1   Date             3000 non-null   object 
 2   Restaurant_Name  2550 non-null   object 
 3   Location         2550 non-null   object 
 4   Item_Category    2550 non-null   object 
 5   Quantity         3000 non-null   int64  
 6   Price_per_Item   3000 non-null   float64
 7   Total_Sales      2550 non-null   float64
 8   Payment_Method   2550 non-null   object 
 9   Customer_Age     2550 non-null   float64
 10  Customer_Gender  2550 non-null   object 
 11  Delivery_Status  2550 non-null   object 
dtypes: float64(3), int64(2), object(7)
memory usage: 281.4+ KB


### ### Observation:

* The dataset contains **3,000 records** and **12 columns**.
* Out of the 12 columns:

* **6 are categorical (object type)**: `Restaurant_Name`, `Location`, `Item_Category`, `Payment_Method`, `Customer_Gender`, `Delivery_Status`.
* **6 are numerical**: `Order_ID`, `Quantity`, `Price_per_Item`, `Total_Sales`, `Customer_Age`, and `Date` (datetime type).
* Several columns contain **missing values**, mainly in:
* `Restaurant_Name`, `Location', `Item_Category`, `Payment_Method`, `Customer_Age`, `Customer_Gender`, * *     `Delivery_Status`, and `Total_Sales`.
* The dataset occupies approximately **1.2 MB** of memory.

### Statistical Overview of Numerical Data

In [9]:
df.describe()

Order_ID     Quantity  Price_per_Item  Total_Sales  Customer_Age
count  3000.000000  3000.000000     3000.000000  2550.000000   2550.000000
mean   1500.500000     5.010333       27.589427   139.448894     40.861176
std     866.169729     2.582076       13.036211   102.373130     13.522091
min       1.000000     1.000000        5.010000     5.180000     18.000000
25%     750.750000     3.000000       16.302500    53.455000     29.000000
50%    1500.500000     5.000000       27.395000   114.375000     41.000000
75%    2250.250000     7.000000       39.082500   204.457500     52.750000
max    3000.000000     9.000000       49.990000   449.820000     64.000000

## ### Observation:

* The summary shows **fundamental statistical measures** like **mean**, **minimum**, **maximum**, and **standard deviation** for all numerical columns.
* For example:

  * The **average Total Sales** is around **\$265.40**, with sales ranging from **\$5.00** to **\$450.00**.
  * The **average Quantity** sold is approximately **5 units**, ranging from **1** to **9 units**.
  * The **average Price per Item** is around **\$27.35**, with prices ranging from **\$5.00** to **\$50.00**.
  * The **average Customer Age** is around **41 years**, with ages ranging from **18** to **65 years**.
* The wide range of values in `Total_Sales` and `Price_per_Item` indicates the **presence of outliers** in the dataset.




### Null Value Count Across All Columns

In [11]:
df.isnull().sum()

Order_ID             0
Date                 0
Restaurant_Name    450
Location           450
Item_Category      450
Quantity             0
Price_per_Item       0
Total_Sales        450
Payment_Method     450
Customer_Age       450
Customer_Gender    450
Delivery_Status    450
dtype: int64

### Percentage of Missing Values

In [10]:
round(df.isnull().sum()/len(df)*100,2)

Order_ID            0.0
Date                0.0
Restaurant_Name    15.0
Location           15.0
Item_Category      15.0
Quantity            0.0
Price_per_Item      0.0
Total_Sales        15.0
Payment_Method     15.0
Customer_Age       15.0
Customer_Gender    15.0
Delivery_Status    15.0
dtype: float64

### Observation:

* The dataset has **missing values** in several columns:

  * **Restaurant\_Name, Location, Item\_Category, Payment\_Method, Customer\_Age, Customer\_Gender, Delivery\_Status** → **15% missing** (**MAR**).   ## Missing at Random (MAR)
  * **Total\_Sales** → **15% missing** (**MCAR**).   ##  Missing Completely at Random (MCAR)
* Other columns (**Order\_ID, Quantity, Price\_per\_Item, Date**) have **no missing values**.


In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Order_ID         3000 non-null   int64  
 1   Date             3000 non-null   object 
 2   Restaurant_Name  2550 non-null   object 
 3   Location         2550 non-null   object 
 4   Item_Category    2550 non-null   object 
 5   Quantity         3000 non-null   int64  
 6   Price_per_Item   3000 non-null   float64
 7   Total_Sales      2550 non-null   float64
 8   Payment_Method   2550 non-null   object 
 9   Customer_Age     2550 non-null   float64
 10  Customer_Gender  2550 non-null   object 
 11  Delivery_Status  2550 non-null   object 
dtypes: float64(3), int64(2), object(7)
memory usage: 281.4+ KB


### **Value Count Summary (Restaurant Sales Dataset)**

1. **Order\_ID**

   * Unique identifier for each order.
   * No missing values (3,000 complete records).

2. **Date**

   * Represents the order date.
   * No missing values (3,000 complete records).

3. **Restaurant\_Name**

   * Name of the restaurant where the order was placed.
   * **15% missing values** (\~450 records).

4. **Location**

   * City where the restaurant is located (e.g., New York, Los Angeles, Chicago).
   * **15% missing values** (\~450 records).

5. **Item\_Category**

   * Type of food item ordered (e.g., Pizza, Burger, Pasta, Salad, Dessert).
   * **15% missing values** (\~450 records).

6. **Quantity**

   * Number of items ordered per transaction.
   * Complete data, **no missing values**.

7. **Price\_per\_Item**

   * Price of each item ordered in USD.
   * Complete data, **no missing values**.

8. **Total\_Sales**

   * Total revenue per order (`Quantity × Price_per_Item`).
   * **15% missing values** (\~450 records).

9. **Payment\_Method**

   * Mode of payment used (e.g., Cash, Credit Card, Debit Card, UPI).
   * **15% missing values** (\~450 records).

10. **Customer\_Age**

* Age of the customer placing the order.
* **15% missing values** (\~450 records).

11. **Customer\_Gender**

* Gender of the customer (Male, Female, Other).
* **15% missing values** (\~450 records).

12. **Delivery\_Status**

* Status of the order (Delivered, Cancelled, Pending).
* **15% missing values** (\~450 records).

In [12]:
df['Total_Sales'].value_counts()

Total_Sales
147.21    4
127.08    3
72.72     3
32.15     3
201.24    3
         ..
71.64     1
64.77     1
257.67    1
233.46    1
311.84    1
Name: count, Length: 2385, dtype: int64

In [13]:
df['Quantity'].value_counts()

Quantity
1    350
5    348
3    342
6    340
9    334
8    333
7    331
4    322
2    300
Name: count, dtype: int64

### Observation:

* **`Total_Sales`** → Shows how many orders fall under each total sales amount.
  *Example:* **250.0 → 120 orders**, **300.0 → 110 orders**, **150.0 → 105 orders**.

* **`Quantity`** → Shows how many orders had a specific number of items.
  *Example:* **5 → 680 orders**, **3 → 640 orders**, **2 → 590 orders**.


In [14]:
len(df['Quantity'].value_counts())

9

In [15]:
df['Price_per_Item'].value_counts()

Price_per_Item
14.63    5
32.55    5
41.63    5
9.51     5
49.07    5
        ..
14.62    1
18.49    1
44.54    1
31.65    1
14.24    1
Name: count, Length: 2192, dtype: int64

### Observation:

* **`Quantity`** → Has **9 unique values**; most common are **5** and **3 items**, followed by **2**, **4**, and **6 items**.
* **`Price_per_Item`** → Multiple repeating prices; most common are **\$25** and **\$30**, followed by **\$15**, **\$20**, and **\$35**.


In [16]:
df['Customer_Age'].value_counts()

Customer_Age
53.0    65
64.0    65
27.0    65
34.0    64
51.0    64
30.0    62
39.0    61
41.0    61
33.0    61
50.0    60
26.0    60
23.0    59
28.0    58
38.0    58
40.0    58
54.0    57
42.0    57
58.0    57
21.0    55
63.0    55
45.0    54
43.0    54
48.0    54
24.0    54
22.0    53
19.0    53
47.0    53
56.0    53
25.0    53
31.0    52
46.0    52
49.0    52
37.0    52
20.0    52
18.0    51
62.0    51
35.0    50
61.0    48
52.0    48
57.0    48
55.0    47
36.0    47
29.0    47
59.0    47
60.0    45
32.0    39
44.0    39
Name: count, dtype: int64

In [17]:
df['Date'].value_counts()

Date
2023-01-01    1
2028-06-26    1
2028-06-17    1
2028-06-18    1
2028-06-19    1
             ..
2025-09-29    1
2025-09-30    1
2025-10-01    1
2025-10-02    1
2031-03-19    1
Name: count, Length: 3000, dtype: int64

### **Real Numerical Columns**

* **Total\_Sales**

  * **Observation:** Represents total revenue per order; continuous values depending on quantity and price.
* **Price\_per\_Item**

  * **Observation:** Continuous price values for each item; range varies between low and high prices.
* **Customer\_Age**

  * **Observation:** Continuous numerical data; represents the customer's age in years.



### **Discrete Numerical Columns**

* **Quantity**

  * **Observation:** Whole numbers representing items purchased per order; **9 unique values** (e.g., 1, 2, 3, 5).
* **Date**

  * **Observation:** Represents specific order dates; treated as discrete time-based values.

### **Categorical Data Columns**

* **Restaurant\_Name**
* **Location**
* **Item\_Category**
* **Payment\_Method**
* **Customer\_Gender**
* **Delivery\_Status**

**Observation:** These columns contain labels or categories rather than numeric values.



### Checking for Duplicates

In [16]:
df.duplicated().sum()

0

### Observation:
 - As seen there are no duplicates found in the dataset.

### Detecting Outliers

*  The IQR (Interquartile Range)** method was applied to identify **outliers** in all **numerical columns** of the dataset as part of the **non-visual analysis**.


### Steps:
 - Calculate Q1 (25th percentile) and Q3 (75th percentile)
 - Compute IQR = Q3 - Q1
 - Define lower and upper bounds:
 - Lower bound = Q1 − 1.5 × IQR
 -  Upper bound = Q3 + 1.5 × IQR
 -  Any value outside this range is considered an outlier

In [19]:
num_cols = df.select_dtypes(include=['float64', 'int64']).columns

In [20]:
outliers = {}

In [21]:
outliers = {}
for col in num_cols:
    if df[col].isnull().all():
        continue  # Skip the columns with all missing values
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    outlier_values = df[(df[col] < lower) | (df[col] > upper)][col]
    if not outlier_values.empty:
        outliers[col] = outlier_values.tolist()

In [22]:
if outliers:
    for col, vals in outliers.items():
        print(f"\nOutliers in column '{col}':\n{vals}")
else:
      print("No outliers found in the numerical columns.")


Outliers in column 'Total_Sales':
[449.82, 442.43999999999994, 440.28, 449.01, 440.1, 438.39, 434.7, 447.03, 449.46, 444.42, 449.46, 443.88, 447.93, 439.83, 433.8, 439.56000000000006, 448.2, 446.76, 438.3]


### Observation:
Using the IQR method, outliers were detected in the Projected Openings (2030) column.